src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,561 @@
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+}
+
+#[allow(dead_code)]
+fn get_obj_binary_len(data: &BTreeMap<String, Json>) -> usize {
+    let element_count = data.len();
+    let key_entries_len = element_count * LENGTH_KEY_ENTRY;
+    let value_entries_len = element_count * LENGTH_VALUE_ENTRY;
+    let mut keys_len = 0;
+    let mut values_len = 0;
+    for (k, v) in data {
+        keys_len += k.len();
+        if v.get_type_code() != TYPE_CODE_LITERAL {
+            values_len += v.binary_len() - LENGTH_TYPE;
+        }
+    }
+    // object: element-count size key-entry* value-entry* key* value*
+    LENGTH_U32 + LENGTH_U32 + key_entries_len + value_entries_len + keys_len + values_len
+}
+
+#[allow(dead_code)]
+fn get_array_binary_len(data: &[Json]) -> usize {
+    let element_count = data.len();
+    let value_entries_len = element_count * LENGTH_VALUE_ENTRY;
+    let mut values_len = 0;
+    for v in data {
+        if v.get_type_code() != TYPE_CODE_LITERAL {
+            values_len += v.binary_len() - LENGTH_TYPE;
+        }
+    }
+    // array ::= element-count size value-entry* value*
+    LENGTH_U32 + LENGTH_U32 + value_entries_len + values_len
+}
+
+#[allow(dead_code)]
+fn get_str_binary_len(data: &str) -> usize {
+    let len = data.as_bytes().len();
+    get_var_u64_blen(len as u64) + len
+}
+
+#[allow(dead_code)]
+fn get_var_u64_blen(mut v: u64) -> usize {
+    let mut len = 1;
+    while v >= 0x80 {
+        v >>= 7;
+        len += 1;
+    }
+    len
+}
+
+impl Serialize for Json {
+    fn serialize<S>(&self, serializer: S) -> SResult<S::Ok, S::Error>
+        where S: Serializer
+    {
+        match *self {
+            Json::Literal(l) => {
+                match l {
+                    JSON_LITERAL_NIL => serializer.serialize_none(),
+                    JSON_LITERAL_TRUE => serializer.serialize_bool(true),
+                    _ => serializer.serialize_bool(false),
+                }
+            }
+            Json::String(ref s) => serializer.serialize_str(s),
+            Json::Object(ref obj) => {
+                let mut map = try!(serializer.serialize_map(Some(obj.len())));
+                for (k, v) in obj {
+                    try!(map.serialize_entry(k, v));
+                }
+                map.end()
+            }
+            Json::Array(ref array) => {
+                let mut tup = try!(serializer.serialize_tuple(array.len()));
+                for item in array {
+                    try!(tup.serialize_element(item));
+                }
+                tup.end()
+            }
+            Json::Double(d) => serializer.serialize_f64(d),
+            Json::I64(d) => serializer.serialize_i64(d),
+        }
+    }
+}
+
+struct JsonVisitor;
+impl<'de> Visitor<'de> for JsonVisitor {
+    type Value = Json;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "may be any string")
```

Comment:
`expecting` is used  in error messages, and since the string we expect may be one of `string`,`int`,`array`, `object`,`boolean` and `null`, so I use  `"may be any string"` here. 
https://docs.serde.rs/serde/de/trait.Visitor.html#tymethod.expecting @BusyJay